File: dotnet/examples/Pactman.NonprofitCheckPlus.Examples/Cases/Ex03IdentityLookup.cs
using System;
using System.Text.Json;
using System.Threading.Tasks;

namespace Pactman.NonprofitCheckPlus.Examples.Cases;

/// <summary>
/// EX-03 — Identity lookup.
/// </summary>
/// <remarks>
/// EIN, name, AKA and Pactman profile URL, plus the raw envelope alongside the typed model.
/// </remarks>
public sealed class Ex03IdentityLookup : IExample
{
    /// <inheritdoc />
    public string Id => "ex-03";

    /// <inheritdoc />
    public string Title => "Identity fields, and the raw envelope alongside the typed model.";

    /// <inheritdoc />
    public async Task RunAsync(string[] args)
    {
        using var context = ExampleContext.Live();

        var result = await context.Client.Nonprofits.CheckAsync(
            ExampleContext.Argument(args, 0, "41-1787097"));

        if (result.Nonprofit is null)
        {
            Output.Note("The API returned no record for that EIN. That is an answer, not an error.");

            return;
        }

        var organization = result.Nonprofit;

        Output.Heading("Identity");
        Output.DisplayField(organization, "ein");
        Output.DisplayField(organization, "organization_name");
        Output.DisplayField(organization, "organization_name_aka");
        Output.DisplayField(organization, "pactman_org_url");
        Output.DisplayField(organization, "organization_info_last_modified");

        Output.Heading("The same fields, typed");
        Output.Field("Ein", organization.Ein);
        Output.Field("OrganizationName", organization.OrganizationName);
        Output.Field("OrganizationNameAka", organization.OrganizationNameAka);

        Output.Heading("The envelope around it");
        Output.Field("Status", result.Status);
        Output.Field("RequestId", result.RequestId);
        Output.Field("TimeTakenMs", result.TimeTakenMs);
        Output.Field("CheckCount", result.CheckCount);
        Output.Field("Errors", result.Errors.Count);

        // The parsed body is kept whole, so anything not typed above is still reachable.
        if (result.Raw.IsJson)
        {
            Output.Field("raw.code", result.Raw.Envelope.GetProperty("code").GetRawText());
            Output.Field("raw.message", result.Raw.Envelope.GetProperty("message").GetRawText());
        }

        Output.Heading("Fields the API returned");
        Output.Field("count", organization.Count);
        Output.Field("names", string.Join(", ", organization.FieldNames));

        Output.Note(
            "Typed properties and wire names read the same store. Nothing the API sent is\n"
            + "dropped on the way through.");
    }
}

File: ruby/examples/ex_03_identity_lookup.rb
# frozen_string_literal: true

# EX-03 — Basic nonprofit identity lookup.
#
# Retrieves an organization and reads its identity fields. The model and the
# untouched response body are both available; neither replaces the other.
#
# Run:  PACTMAN_API_KEY=... bundle exec ruby examples/ex_03_identity_lookup.rb [EIN]

require_relative "lib/client"
require_relative "lib/fixture_api"
require_relative "lib/print"

include Print

client = ExampleClient.create_client
ein = ARGV.fetch(0, FixtureApi::EINS[:public_charity])

result = client.nonprofits.check(ein)
nonprofit = result.nonprofit

if nonprofit.nil?
  puts "No record for EIN #{ein}."
  exit 0
end

heading "Identity"
field "ein", nonprofit.ein
field "organization_name", nonprofit.organization_name
field "organization_name_aka", returned(nonprofit, :organization_name_aka)
field "pactman_org_url", nonprofit.pactman_org_url

# `organization_name_aka` is frequently null. That is "the API has no alternate
# name on file", not "the organization has no alternate name".

heading "Response metadata"
field "status", result.status
field "request_id", result.request_id
field "time_taken_ms", result.time_taken_ms
field "check_count", result.check_count

# The model is a view over the envelope, not a replacement for it. `raw` is
# exactly what the server sent, including anything not declared on the model.
heading "Raw envelope"
field 'raw["code"]', result.raw["code"]
field 'raw["message"]', result.raw["message"]
field 'raw["data"]["ein"]', result.raw.dig("data", "ein")
field "fields returned", nonprofit.keys.size

note "A returned profile URL means Pactman holds a page for the organization. It is\n" \
     "not an endorsement, and not a statement about tax-exempt status."
